// controller/index.js

function Controller() {
    this.controllersList = {};
    this.urlManager = {};
    this.db = {};

}

function toControllerName(str) {
    return str[0].toLocaleUpperCase() + str.substr(1) + "Controller";
}
function toActionName(str) {
    return "action" + str[0].toLocaleUpperCase() + str.substr(1);
}

Controller.prototype.createWebApplication = function(obj){
    this.urlManager = obj.urlManager;
    this.db = obj.db;

    return this;
};

Controller.prototype.start = function(app){
    var self = this;

    this.urlManager.forEach(function(urlconf){
        var url = urlconf.url;
        var method = urlconf.method;
        var path = urlconf.path;

        app[method](url, function (req, res) {
            var controller = toControllerName(path[0]);
            var action = toActionName(path[1] || "index");

            var mysql = require('mysql');
            var dbConnrction = mysql.createConnection(self.db);

            if (!self.controllersList[controller] || !self.controllersList[controller][action]) {
                res.render('404', {});
            } else {
                self.controllersList[controller][action](req, res, dbConnrction);
            }

        });
    });

};

Controller.prototype.defineController = function(controllerName, contstructor){
    this.controllersList[controllerName] = contstructor;
};


var inst = new Controller();
module.exports = inst;
